File: app/frontend/src/components/chatui/metricsTracker.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
//
// Metrics measurement approach adapted from the vLLM project (Apache-2.0):
// https://github.com/vllm-project/vllm/blob/main/vllm/benchmarks/lib/endpoint_request_func.py
// SPDX-FileCopyrightText: Copyright contributors to the vLLM project

/**
 * Metrics Tracker for Inference Performance
 *
 * Tracks TTFT, ITL, and TPOT following vLLM's measurement approach:
 * - TTFT: time from request start to first content chunk
 * - ITL: list of intervals between consecutive content chunks (ms)
 * - TPOT: mean of ITL
 * - Network latency: client_ttft - backend_ttft
 */

import type { TokenTimestamp, InferenceStats, ProgressiveStats, HardwareMetrics } from "./types";
import type { GpuBenchmark } from "./benchmarkData";

export class InferenceMetricsTracker {
  // Timing measurements
  private requestStartTime: number = 0;
  private firstTokenTime: number | undefined;
  private firstContentTokenTime: number | undefined; // first non-thinking token (true TTFT)
  private mostRecentTokenTime: number | undefined;   // vLLM-style: advances per content chunk
  private itl: number[] = [];                        // inter-token latencies in ms
  private tokenTimestamps: TokenTimestamp[] = [];

  // Token counters
  private lastTokenCount: number = 0;

  // Thinking/reasoning tracking
  private thinkingStartTime: number | undefined;
  private thinkingEndTime: number | undefined;
  private reasoningTokenCount: number = 0;

  constructor() {
    this.reset();
  }

  /**
   * Reset all metrics for a new inference request
   */
  reset(): void {
    this.requestStartTime = performance.now();
    this.firstTokenTime = undefined;
    this.firstContentTokenTime = undefined;
    this.mostRecentTokenTime = undefined;
    this.itl = [];
    this.tokenTimestamps = [];
    this.lastTokenCount = 0;
    this.thinkingStartTime = undefined;
    this.thinkingEndTime = undefined;
    this.reasoningTokenCount = 0;
  }

  /**
   * Record arrival of a thinking/reasoning chunk.
   * Does not affect ITL or TTFT — thinking time is tracked separately.
   */
  recordThinkingToken(): void {
    const now = performance.now();
    if (!this.thinkingStartTime) {
      this.thinkingStartTime = now;
      if (!this.firstTokenTime) this.firstTokenTime = now;
    }
    this.reasoningTokenCount++;
  }

  /**
   * Record arrival of a content chunk, tracking ITL exactly as vLLM does.
   * Call this once per content delta (text) as it arrives.
   */
  recordContentToken(): void {
    const now = performance.now();
    // Close thinking window on first content token
    if (this.thinkingStartTime && !this.thinkingEndTime) {
      this.thinkingEndTime = now;
    }
    if (!this.firstContentTokenTime) {
      this.firstContentTokenTime = now;
      if (!this.firstTokenTime) this.firstTokenTime = now;
      console.log(`[Metrics] First content token at ${(now - this.requestStartTime).toFixed(2)}ms`);
      // No ITL for first content token
    } else if (this.mostRecentTokenTime !== undefined) {
      this.itl.push(now - this.mostRecentTokenTime);
    }
    this.mostRecentTokenTime = now;
  }

  /**
   * @deprecated Use recordContentToken() for accurate ITL tracking.
   */
  recordFirstToken(): void {
    this.recordContentToken();
  }

  /**
   * Record usage data from a streaming chunk
   * @param usage - Usage object from streaming response containing token counts
   */
  recordUsage(usage: {
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
  }): void {
    const completionTokens = usage.completion_tokens || 0;

    // Only record if we have tokens and it's a new count
    if (completionTokens > 0 && completionTokens > this.lastTokenCount) {
      const timestamp = performance.now();

      this.tokenTimestamps.push({
        count: completionTokens,
        timestamp: timestamp,
        promptTokens: usage.prompt_tokens || 0,
        totalTokens: usage.total_tokens || 0,
      });

      this.lastTokenCount = completionTokens;

      console.log(`[Metrics] Token #${completionTokens} at ${(timestamp - this.requestStartTime).toFixed(2)}ms`);
    }
  }

  /**
   * Calculate progressive statistics for live display
   * @returns Current progressive stats or null if no data available
   */
  getProgressiveStats(): ProgressiveStats | null {
    if (this.tokenTimestamps.length === 0) {
      return null;
    }

    const latestTimestamp = this.tokenTimestamps[this.tokenTimestamps.length - 1];
    const elapsedSeconds = (latestTimestamp.timestamp - this.requestStartTime) / 1000;
    const tokensGenerated = latestTimestamp.count;
    const tokensPerSecond = tokensGenerated / elapsedSeconds;

    return {
      tokensGenerated,
      tokensPerSecond,
      elapsedSeconds,
    };
  }

  /**
   * Finalize metrics and attach client-side measurements to backend stats.
   * @param backendStats - Stats received from backend
   * @returns Enhanced stats with client-side measurements
   */
  finalizeStats(backendStats: InferenceStats): InferenceStats {
    const enhancedStats = { ...backendStats };

    // Calculate client-side TTFT (use first CONTENT token for thinking models)
    const ttftAnchor = this.firstContentTokenTime ?? this.firstTokenTime;
    if (ttftAnchor) {
      const clientTtftMs = ttftAnchor - this.requestStartTime;
      const backendTtftMs = (backendStats.user_ttft_s || 0) * 1000;

      enhancedStats.client_ttft_ms = clientTtftMs;
      enhancedStats.network_latency_ms = Math.max(0, clientTtftMs - backendTtftMs);

      console.log(`[Metrics] Client TTFT: ${clientTtftMs.toFixed(2)}ms`);
      console.log(`[Metrics] Backend TTFT: ${backendTtftMs.toFixed(2)}ms`);
      console.log(`[Metrics] Network Latency: ${enhancedStats.network_latency_ms.toFixed(2)}ms`);
    }

    // Attach client-side ITL list (ms)
    if (this.itl.length > 0) {
      enhancedStats.itl = this.itl;
      console.log(`[Metrics] ITL samples: ${this.itl.length}, mean: ${(this.itl.reduce((a, b) => a + b, 0) / this.itl.length).toFixed(2)}ms`);
    }

    // Attach token timestamps for per-token analysis
    if (this.tokenTimestamps.length > 0) {
      enhancedStats.token_timestamps = this.tokenTimestamps;
      console.log(`[Metrics] Captured ${this.tokenTimestamps.length} token timestamps`);
    }

    // Attach thinking/reasoning metrics (client-measured)
    if (this.reasoningTokenCount > 0) {
      enhancedStats.reasoning_tokens = this.reasoningTokenCount;
    }
    if (this.thinkingStartTime && this.thinkingEndTime) {
      enhancedStats.thinking_duration_ms = this.thinkingEndTime - this.thinkingStartTime;
      console.log(`[Metrics] Thinking duration: ${enhancedStats.thinking_duration_ms.toFixed(0)}ms, reasoning tokens: ${this.reasoningTokenCount}`);
    }

    return enhancedStats;
  }

  /**
   * Get the request start time
   */
  getRequestStartTime(): number {
    return this.requestStartTime;
  }

  /**
   * Get all recorded token timestamps
   */
  getTokenTimestamps(): TokenTimestamp[] {
    return this.tokenTimestamps;
  }

  /**
   * Check if first token has been received
   */
  hasReceivedFirstToken(): boolean {
    return this.firstTokenTime !== undefined;
  }
}

/**
 * Compute efficiency comparison between TT hardware and GPU baselines.
 */
export interface EfficiencyComparison {
  gpu: string;
  gpu_tps: number;
  gpu_tps_per_watt: number;
  tt_tps: number;
  tt_tps_per_watt: number;
  speed_ratio: number;
  efficiency_ratio: number;
}

export function computeEfficiencyComparisons(
  stats: InferenceStats,
  baselines: GpuBenchmark[],
): EfficiencyComparison[] {
  const tps =
    stats.tps ??
    (typeof stats.user_tpot === "number" && stats.user_tpot > 0
      ? 1 / stats.user_tpot
      : undefined);
  const tpsPerWatt = stats.tps_per_watt;

  if (!tps || !tpsPerWatt) return [];

  return baselines.map((b) => ({
    gpu: b.gpu,
    gpu_tps: b.tps,
    gpu_tps_per_watt: b.tps_per_watt,
    tt_tps: tps,
    tt_tps_per_watt: tpsPerWatt,
    speed_ratio: Math.round((tps / b.tps) * 100) / 100,
    efficiency_ratio: Math.round((tpsPerWatt / b.tps_per_watt) * 10) / 10,
  }));
}

File: app/backend/model_control/metrics_tracker.py
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
#
# Metrics measurement approach adapted from the vLLM project (Apache-2.0):
# https://github.com/vllm-project/vllm/blob/main/vllm/benchmarks/lib/endpoint_request_func.py
# SPDX-FileCopyrightText: Copyright contributors to the vLLM project

"""
Backend Metrics Tracker for Inference Performance

Accurate metrics calculation following vLLM's measurement approach:
- TTFT (Time to First Token): time from request start to first content chunk
- ITL (Inter-Token Latency): list of intervals between consecutive content chunks
- TPOT (Time Per Output Token): mean of ITL
- Token counting and timing
"""

import time
from typing import Optional, Dict, List


class InferenceMetricsTracker:
    """Track inference metrics during streaming, following vLLM's measurement approach."""

    def __init__(self):
        self.start_time: float = time.perf_counter()
        self.first_token_time: Optional[float] = None
        self.most_recent_token_time: Optional[float] = None  # vLLM-style: advances per content chunk
        self.itl: List[float] = []          # inter-token latency list (seconds) — vLLM style
        self.token_times: List[float] = []  # kept for backwards compat / detailed stats
        self.num_tokens: int = 0
        self.prompt_tokens: int = 0
        self.last_token_count: int = 0

    def record_content_token(self) -> None:
        """Record arrival of a content chunk, tracking ITL exactly as vLLM does."""
        current_time = time.perf_counter()
        if self.first_token_time is None:
            self.first_token_time = current_time
        else:
            self.itl.append(current_time - self.most_recent_token_time)
        self.most_recent_token_time = current_time
        self.token_times.append(current_time)
        self.num_tokens += 1
        self.last_token_count = self.num_tokens

    def set_prompt_tokens(self, prompt_tokens: int) -> None:
        """Set prompt token count from usage data"""
        if self.prompt_tokens == 0:
            self.prompt_tokens = prompt_tokens

    def record_token(self, completion_tokens: int, prompt_tokens: int = 0) -> None:
        """
        Record token arrival from usage data (cloud path).

        Args:
            completion_tokens: Current completion token count
            prompt_tokens: Number of prompt tokens (captured once)
        """
        current_time = time.perf_counter()

        # Record first token
        if completion_tokens == 1 and self.first_token_time is None:
            self.first_token_time = current_time
            self.prompt_tokens = prompt_tokens
            self.token_times.append(current_time)
            self.num_tokens = 1
            self.last_token_count = 1

        # Record subsequent tokens
        elif completion_tokens > self.last_token_count:
            # Handle batch arrivals (e.g., token count jumps from 5 to 10)
            tokens_arrived = completion_tokens - self.last_token_count
            for _ in range(tokens_arrived):
                self.token_times.append(current_time)

            self.num_tokens = completion_tokens
            self.last_token_count = completion_tokens

    def get_ttft(self) -> float:
        """Time to first content chunk, in seconds. Returns 0 if no tokens received."""
        if self.first_token_time is None:
            return 0.0
        return self.first_token_time - self.start_time

    def get_tpot(self) -> float:
        """
        Calculate Time Per Output Token as mean of ITL — matches vLLM's approach.

        Returns:
            Mean inter-token latency in seconds, or 0 if insufficient data
        """
        if not self.itl:
            return 0.0
        return sum(self.itl) / len(self.itl)

    def get_accurate_tpot(self) -> float:
        """Median of ITL — more robust to outliers than the mean."""
        if not self.itl:
            return 0.0
        sorted_itl = sorted(self.itl)
        mid = len(sorted_itl) // 2
        if len(sorted_itl) % 2 == 0:
            return (sorted_itl[mid - 1] + sorted_itl[mid]) / 2
        return sorted_itl[mid]

    def get_stats(self) -> Dict:
        """Get final statistics in vLLM-compatible format."""
        return {
            "ttft": self.get_ttft(),
            "tpot": self.get_tpot(),
            "itl": self.itl,              # list of inter-token latencies in seconds
            "tokens_decoded": self.num_tokens,
            "tokens_prefilled": self.prompt_tokens,
            "context_length": self.prompt_tokens + self.num_tokens,
            "total_time": time.perf_counter() - self.start_time,
        }

    def get_detailed_stats(self) -> Dict:
        """Get detailed statistics with ITL percentiles."""
        basic_stats = self.get_stats()

        if self.itl:
            sorted_itl = sorted(self.itl)
            n = len(sorted_itl)
            return {
                **basic_stats,
                "tpot_median": sorted_itl[n // 2],
                "tpot_p95": sorted_itl[int(n * 0.95)],
                "tpot_p99": sorted_itl[int(n * 0.99)],
                "tpot_min": sorted_itl[0],
                "tpot_max": sorted_itl[-1],
            }

        return basic_stats
